File: forecasto-server/src/forecasto/schemas/invoice.py
"""Invoice (fattura attiva) schemas.

Invoices are stored as documents in the per-workspace ``invoices`` collection.
Monetary values are stored as decimal *strings* in the document JSON to preserve
cent precision (JSON has no Decimal); these schemas accept ``Decimal`` on input
and the service stringifies on store.
"""

from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from typing import Any

from pydantic import BaseModel, Field


class InvoiceLineIn(BaseModel):
    id: str | None = None
    code: str | None = None  # article/service code (FatturaPA CodiceArticolo)
    name: str | None = None
    description: str | None = None
    quantity: Decimal = Decimal("1")
    unit_code: str = "C62"
    net_unit_price: Decimal
    discount_percent: Decimal | None = None  # line discount % (ScontoMaggiorazione)
    line_net_amount: Decimal | None = None  # computed from qty×price×(1−disc) if omitted
    vat_rate: Decimal = Decimal("22")
    vat_category: str = "S"  # EN16931 category (S=standard, N=non-taxable, ...)
    natura: str | None = None  # FatturaPA Natura for non-taxable lines (e.g. "N3.5")


class ScadenzaIn(BaseModel):
    id: str | None = None
    due_date: date
    amount: Decimal | None = None  # auto-distributed from the grand total if omitted
    modalita: str = "MP05"  # FatturaPA ModalitaPagamento (MP05 = bank transfer)


class PaymentsIn(BaseModel):
    means_code: str = "30"  # EN16931 payment means (30 = credit transfer)
    esigibilita_iva: str = "I"  # I=immediate, D=deferred, S=split payment
    terms: str | None = None  # free-text terms, e.g. "30/60/90 df fm" (parsed to scadenze)
    scadenze: list[ScadenzaIn] = Field(default_factory=list)


class PaymentTermsParseRequest(BaseModel):
    """Preview the due dates a payment-terms string yields for a given date."""

    text: str
    issue_date: date


class SdiSubmissionRequest(BaseModel):
    """Record a manual SDI submission, optionally with its outcome."""

    outcome: str | None = None  # None | "accepted" | "rejected"


class InvoiceDraftCreate(BaseModel):
    customer_document_id: str | None = None
    type_code: str = "380"  # 380 invoice, 381 credit note
    currency: str = "EUR"
    issue_date: date | None = None
    causale: str | None = None
    lines: list[InvoiceLineIn] = Field(default_factory=list)
    payments: PaymentsIn = Field(default_factory=PaymentsIn)
    fattura_pa_ext: dict[str, Any] | None = None
    extended: dict[str, Any] = Field(default_factory=dict)
    source_order_record_ids: list[str] = Field(default_factory=list)
    intent_letter_id: str | None = None


class InvoiceUpdate(BaseModel):
    customer_document_id: str | None = None
    type_code: str | None = None
    currency: str | None = None
    issue_date: date | None = None
    causale: str | None = None
    lines: list[InvoiceLineIn] | None = None
    payments: PaymentsIn | None = None
    fattura_pa_ext: dict[str, Any] | None = None
    extended: dict[str, Any] | None = None
    intent_letter_id: str | None = None


class InvoiceResponse(BaseModel):
    document_id: str
    status: str
    number: str | None
    data: dict[str, Any]
    created_at: datetime
    updated_at: datetime

File: forecasto-client-web/src/types/invoice.ts
// Monetary values are decimal strings (server stores them as strings to keep
// cent precision). Inputs may pass numbers or strings.

export interface InvoiceLine {
  id?: string
  code?: string | null
  name?: string | null
  description?: string | null
  quantity: string
  unit_code?: string
  net_unit_price: string
  discount_percent?: string | null
  line_net_amount?: string
  vat_rate: string
  vat_category?: string
  natura?: string | null
}

export interface Scadenza {
  id?: string
  due_date: string
  amount?: string
  modalita?: string
  record_id?: string | null
}

export interface Payments {
  means_code: string
  esigibilita_iva: string
  terms?: string | null
  scadenze: Scadenza[]
}

export interface ParsedScadenza {
  days: number
  due_date: string
  end_of_month: boolean
  label: string
}

export interface EInvoiceSummary {
  document_id: string
  standard: string
  filename: string
  generated_at: string
  validation: { ok: boolean; errors: string[]; warnings: string[] }
  transmission: { sent_at: string | null; outcome: string | null } | null
  stale: boolean
}

export interface TaxBreakdownEntry {
  category: string
  rate: string
  taxable_amount: string
  tax_amount: string
  natura?: string
}

export interface Totals {
  line_total: string
  allowance_total: string
  charge_total: string
  tax_basis_total: string
  tax_total: string
  grand_total: string
  prepaid_amount: string
  rounding_amount: string
  due_payable: string
}

export interface InvoiceLifecycle {
  status: string
  created_at?: string
  issued_at?: string | null
  sent_to_client_at?: string | null
  sdi_submitted_at?: string | null
  sdi_outcome?: string | null
  sdi_outcome_at?: string | null
}

export interface InvoiceLinks {
  actual_record_ids: string[]
  einvoice_doc_ids: string[]
  source_order_record_ids: string[]
  credit_note_of: string | null
  intent_letter_id?: string | null
}

export interface InvoiceData {
  kind?: string
  number: string | null
  issue_date: string | null
  type_code: string
  currency: string
  causale?: string | null
  emitter?: Record<string, unknown> | null
  customer_ref?: { document_id: string; vat_id?: string | null } | null
  customer_snapshot?: Record<string, unknown> | null
  lines: InvoiceLine[]
  tax_breakdown: TaxBreakdownEntry[]
  totals: Totals
  payments: Payments
  fattura_pa_ext?: Record<string, unknown>
  extended?: Record<string, unknown>
  lifecycle: InvoiceLifecycle
  links: InvoiceLinks
  sync?: { data_fingerprint?: string; last_synced_at?: string }
}

export interface Invoice {
  document_id: string
  status: string
  number: string | null
  data: InvoiceData
  created_at: string
  updated_at: string
}

// ---- inputs ----

export interface InvoiceLineInput {
  id?: string
  code?: string
  name?: string
  description?: string
  quantity?: string | number
  net_unit_price: string | number
  discount_percent?: string | number | null
  vat_rate?: string | number
  vat_category?: string
  natura?: string | null
}

export interface ScadenzaInput {
  id?: string
  due_date: string
  amount?: string | number | null
  modalita?: string
}

export interface InvoiceDraftCreate {
  customer_document_id?: string | null
  type_code?: string
  currency?: string
  issue_date?: string | null
  causale?: string | null
  lines: InvoiceLineInput[]
  payments?: { means_code?: string; esigibilita_iva?: string; terms?: string | null; scadenze: ScadenzaInput[] }
  fattura_pa_ext?: Record<string, unknown>
  extended?: Record<string, unknown>
  source_order_record_ids?: string[]
  intent_letter_id?: string | null
}

export type InvoiceUpdate = Partial<InvoiceDraftCreate>
